### PR TITLE
Prioritize a profile passed by user over env vars

### PR DESCRIPTION
**Pull request by mikekaminsky**
Please ensure the following before submitting a PR:

 - [X] if suggesting code changes or improvements, [open an issue](https://github.com/cloudyr/aws.signature/issues/new) first

Issue here: https://github.com/cloudyr/aws.signature/issues/54

 - [ ] for all but trivial changes (e.g., typo fixes), add your name to [DESCRIPTION](https://github.com/cloudyr/aws.signature/blob/master/DESCRIPTION)

Is this trivial? Also I'm not sure where to add my name.

 - [X] for all but trivial changes (e.g., typo fixes), documentation your change in [NEWS.md](https://github.com/cloudyr/aws.signature/blob/master/NEWS.md) with a parenthetical reference to the issue number being addressed

 - [X] if changing documentation, edit files in `/R` not `/man` and run `devtools::document()` to update documentation
 - [X] add code or new test files to [`/tests`](https://github.com/cloudyr/aws.signature/tree/master/tests/testthat) for any new functionality or bug fix
 - [X] make sure `R CMD check` runs without error before submitting the PR

Additional Notes:
* I'm assuming we're going to change `CREDENTIALS_FEATURE_FLAG_ON` which has a bad name in favor of some other options-strategy -- waiting for feedback on what to do there
* This code is a little bit gross, and this whole function could probably benefit from a refactor -- I did _**not**_ do that in favor of duplicating the code following the existing style
* Locally two tests were failing locally before making any changes -- those same two tests are still failing but otherwise tests / `R CMD check` looks good. (I think this has to do with the some of the tests looking at my local `~/.aws/config` file)

### CODE REVIEW

File: R/locate_credentials.R [hunk 1]
```diff
@@ -75,12 +82,47 @@ function(
...
+        } else if (file.exists(file) || file.exists(default_credentials_file())) {
+            if (file.exists(file)) {
+                cred <- read_credentials(file = file)
+            } else {
+                ## otherwise, default to default location
+                cred <- read_credentials(file = default_credentials_file())
+            }
+        if (profile %in% names(cred)) {
+            cred <- cred[[profile]]
+                cred <- cred[[profile]]
```
[the remark this answers is not in the document]
Comment (mikekaminsky) [reply R/locate_credentials.R]: Thanks! Fixed.

File: R/locate_credentials.R [hunk 2]
```diff
@@ -45,10 +45,17 @@ function(
...
-  profile = Sys.getenv("AWS_PROFILE", "default"),
+  profile = NULL,
   default_region = getOption("cloudyr.aws.default_region", "us-east-1"),
   verbose = getOption("verbose", FALSE)
 ) {
+
+    profile_set <- TRUE
+    if (is.null(profile)){
+      profile_set <- FALSE
+      profile <- Sys.getenv("AWS_PROFILE", "default")
```
[the remark this answers is not in the document]
Comment (mikekaminsky) [reply R/locate_credentials.R:56]: I'm not sure if you're commenting on the naming convention here or something else -- assuming yes, i changed this to `profile_is_passed` -- if you mean something else let me know!

File: R/locate_credentials.R [hunk 3]
```diff
@@ -179,14 +221,14 @@ function(
...
         cred <- read_credentials(file.path(".aws", "credentials"))
-        if (profile %in% names(cred)) {
-            cred <- cred[[profile]]
-        } else {
-            cred <- cred[["default"]]
-            if (isTRUE(verbose)) {
-                warning(sprintf("Requested profile '%s' not found in file. Using 'default' profile.", profile))
-            }
-        }
+         if (profile %in% names(cred)) {
```
[the remark this answers is not in the document]
Comment (mikekaminsky) [reply R/locate_credentials.R]: Good catch -- my default settings for spacing are different than yours so was doing a bunch of stuff manually. Fixed now

[the remark this answers is not in the document]
Comment (mikekaminsky) [reply R/locate_credentials.R]: (I think)

**Comment by mikekaminsky** [comment]
> For the credentials finding, I wonder if it makes sense to extract out that credentials file section of locate_credentials into another method, as it's currently a complicated mess of if statements (which I know is not your fault), and then have locate_credentials pull the profile out if specified?

I think something like this is a good idea, but I'm not sure I am the best person to implement it -- what i'd probably do is pull out all of the big `if` blocks we currently have and turn them into more isolated functions so that `locate_credentials` is at least easier to read (and it's clear in the code the search-path of first passed arguments, then profile, then env vars, then default, etc)

> Also, if a user explicitly passes a profile name and we can't find it, should we error at that point?

I thought about this a bit and I don't _think_ so -- though maybe we warn? If the credentials that end up getting selected are invalid the error will happen downstream, but I think falling back to the default profile is fine behavior.

Not at all committed to that so if you have strong feelings 100% can be over-ridden.

File: R/locate_credentials.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (jon-mago) [reply R/locate_credentials.R:56]: Ah, I meant that you're doing this check for the `AWS_PROFILE` after you've set `profile_set` to `FALSE`.  But if I've set profile in an environment variable, shouldn't that count as a 'set profile'?  It's indicating a definite choice.

[the remark this answers is not in the document]
Comment (mikekaminsky) [reply R/locate_credentials.R:56]: Mmm good point! Happy to implement this -- what happens if the profile_name environment variable and the keys conflict? 

I think that if there are environment variable keys we take those, then we look for the environment variable profile name. Does that sound right?

[the remark this answers is not in the document]
Comment (jon-mago) [reply R/locate_credentials.R:56]: What does boto3 do in that case?  I feel like we should be consistent

[the remark this answers is not in the document]
Comment (mikekaminsky) [reply R/locate_credentials.R:56]: Cool I'll take a look and try to mimic boto3 behavior. Probably this weekend.

**Comment by jon-mago** [comment]
> I think something like this is a good idea, but I'm not sure I am the best person to implement it -- what i'd probably do is pull out all of the big `if` blocks we currently have and turn them into more isolated functions so that `locate_credentials` is at least easier to read (and it's clear in the code the search-path of first passed arguments, then profile, then env vars, then default, etc)
> 

That's how I'd do it, yes.  Though that might be something I tackle.  I'm fine with considering that out of scope for this PR.

> > Also, if a user explicitly passes a profile name and we can't find it, should we error at that point?
> 
> I thought about this a bit and I don't _think_ so -- though maybe we warn? If the credentials that end up getting selected are invalid the error will happen downstream, but I think falling back to the default profile is fine behavior.
> 
> Not at all committed to that so if you have strong feelings 100% can be over-ridden.

I've had a bit more of a think, and I think that erroring might be the correct response.  Some aws resources are only identified by name within the account, not by a full ARN with account number.  So picking default credentials could run operations intended for a test environment in a prod environment, or visa versa.

**Comment by mikekaminsky** [comment]
That's compelling! I'll make that change as well this weekend.

File: R/locate_credentials.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (mikekaminsky) [reply R/locate_credentials.R:56]: Okay here's the `boto3` behavior -- it prioritizes the `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` variables over the `AWS_PROFILE` variable

```python
import boto3
import os

os.environ["AWS_PROFILE"] = "woody"
os.environ["AWS_ACCESS_KEY_ID"] = "TOINFINITY"
os.environ["AWS_SECRET_ACCESS_KEY"] = "ANDBEYOND"
session = boto3.Session()
credentials = session.get_credentials()
credentials = credentials.get_frozen_credentials()
print(credentials)
# prints "TO INFINITY" and "AND BEYOND"

os.environ["AWS_PROFILE"] = "woody"
os.environ["AWS_ACCESS_KEY_ID"] = ""
os.environ["AWS_SECRET_ACCESS_KEY"] = ""
session = boto3.Session()
credentials = session.get_credentials()
credentials = credentials.get_frozen_credentials()
print(credentials)
# prints "HOWDY" and "PARTNER"
```

**Comment by mikekaminsky** [comment]
@jon-mago I went ahead and made these changes. https://github.com/cloudyr/aws.signature/pull/57/commits/6688428a68a18f6b76081d644baa87d5e94c9338 adds two features and includes the refactor we briefly discussed above.

1. `locate_credentials()` errors on unknown profile
1. `locate_credentials()` prioritizes env var keys if there is a conflicting env profile var

I think the tests do a good job of demonstrating the feature, but given the refactor it might be a good idea to check that the test coverage for the rest of the feature is adequate -- I don't _think_ I broke anything but there's a number of places this could have gone wrong.

Also I'm curious if you have a better pattern for doing this: 

```
    # check for user-supplied values
    user_supplied_values <- check_for_user_supplied_values(key, secret, region, session_token, default_region, verbose)
    if (!is.null(user_supplied_values)){
      # early return
      return(user_supplied_values)
    }
```
It feels sort of clumsy but also seems to work well enough and is reasonably clear so 🤷‍♀️

**Comment by jon-mago** [comment]
Hi

Sorry it's taken me a while to get back to you on this, it's been a week.

I see what you mean about it feeling sort of clumsy, but I also can't think of a better way and the overall clarity of the method is now much improved, so thank you!

Having a look at the coverage, it's mostly the ec2 creds which are not tested, and thats not really been changed.

I think this is now ready to merge, so I'm likely to do that this weekend

Thanks once more!

**Comment by mikekaminsky** [comment]
Woo! Glad to hear it. Thanks for the feedback on this -- I agree it's much improved. Onward and upward with cloudyr! :)

**Pull request merged**